astropy/units/quantity_helper/function_helpers.py

```diff
@@ -1103,3 +1112,120 @@ def unstructured_to_structured(arr, dtype, *args, **kwargs):
     target_unit = StructuredUnit(_build_structured_unit(dtype, arr.unit))
 
     return (arr.to_value(arr.unit), dtype) + args, kwargs, target_unit, None
+
+
+def _izip_units_flat(iterable: Iterable[StructuredUnit | Unit] | StructuredUnit) -> Iterator:
+    """Returns an iterator of collapsing any nested unit structure.
+
+    Parameters
+    ----------
+    iterable : Iterable[StructuredUnit | Unit] or StructuredUnit
+        A structured unit or iterable thereof.
+
+    Yields
+    ------
+    (str, unit)
+    """
+    from astropy.units import StructuredUnit
+
+    # Make Structured unit.
+    units = iterable if isinstance(iterable, StructuredUnit) else StructuredUnit(iterable)
+
+    # Yield from structured unit.
+    for k, v in units.items():
+        if isinstance(v, StructuredUnit):
+            yield from _izip_units_flat(v)
+        else:
+            yield (k, v)
+
+
+def merge_units(*units: UnitBase | StructuredUnit | None) -> StructuredUnit:
+    """Merge a sequence of units into a structured unit.
+
+    .. todo::
+
+        Either make this function public or incorporate into ``StructuredUnit``.
+
+    Parameters
+    ----------
+    *units : `astropy.units.UnitBase` or `astropy.units.StructuredUnit`
+        The units to merge into one structured unit.
+
+    Returns
+    -------
+    `astropy.units.StructuredUnit`
+        Structured unit of constituent unit.
+    """
+    from astropy.units import StructuredUnit, UnitBase
+
+    # filter units, excluding None
+    actual_units = tuple(unit for unit in units if isinstance(unit, (UnitBase, StructuredUnit)))
+    flat = tuple(_izip_units_flat(StructuredUnit(actual_units)))
+    return StructuredUnit(tuple(uu for _, uu in flat), names=tuple(n for n, _ in flat))
+
+
+@function_helper(helps=rfn.merge_arrays)
+def merge_arrays(
+    seqarrays: Quantity | Sequence[Quantity],
+    fill_value: float = -1,
+    flatten: bool = False,
+    usemask: bool = False,
+    asrecarray: bool = False,
+) -> tuple[tuple[Any, ...], dict[str, Any], StructuredUnit, None]:
+    """Merge structured Quantities field by field.
```

Comment:
For `Quantity`, I have not really gotten there, but the idea is to make it like for `utils.masked.function_helpers`, that docstrings are used only when there is any surprise at what the function does relative to the numpy version -- that way they can be selected to by typeset -- and that the docstrings just state it is like the numpy equivalent except for XXX (see https://docs.astropy.org/en/latest/api/astropy.utils.masked.function_helpers.broadcast_arrays.html#astropy.utils.masked.function_helpers.broadcast_arrays)

This has the large advantage that there is no need to keep the numpy and helper docstrings in sync...

Reply:
That is easier. 

☑